agent/grpc/resolver/resolver.go

```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+// Run periodically reshuffles the order of server addresses  within the
+// resolvers to ensure the load is balanced across servers.
+//
+// TODO: this looks very similar to agent/router.Manager.Start, which is the
+// only other caller of ComputeRebalanceTimer. Are the values passed to these
+// two functions different enough that we need separate goroutines to rebalance?
+// or could we have a single thing handle the timers, and call both rebalance
+// functions?
+func (s *ServerResolverBuilder) Run(ctx context.Context) {
+	// Compute the rebalance timer based on the number of local servers and nodes.
+	rebalanceDuration := router.ComputeRebalanceTimer(s.serversInDC(s.datacenter), s.nodes.NumNodes())
+	timer := time.NewTimer(rebalanceDuration)
+
+	for {
+		select {
+		case <-timer.C:
+			s.rebalanceResolvers()
+
+			// Re-compute the wait duration.
+			newTimerDuration := router.ComputeRebalanceTimer(s.serversInDC(s.datacenter), s.nodes.NumNodes())
+			timer.Reset(newTimerDuration)
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		}
+	}
+}
+
+// rebalanceResolvers shuffles the server list for resolvers in all datacenters.
+func (s *ServerResolverBuilder) rebalanceResolvers() {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	for _, resolver := range s.resolvers {
+		// Shuffle the list of addresses using the last list given to the resolver.
+		resolver.addrLock.Lock()
+		addrs := resolver.addrs
+		rand.Shuffle(len(addrs), func(i, j int) {
+			addrs[i], addrs[j] = addrs[j], addrs[i]
+		})
+		// Pass the shuffled list to the resolver.
+		resolver.updateAddrsLocked(addrs)
+		resolver.addrLock.Unlock()
+	}
+}
+
+// serversInDC returns the number of servers in the given datacenter.
+func (s *ServerResolverBuilder) serversInDC(dc string) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	var serverCount int
+	for _, server := range s.servers {
+		if server.Datacenter == dc {
+			serverCount++
+		}
+	}
+
+	return serverCount
+}
+
+// ServerForAddr returns server metadata for a server with the specified address.
+func (s *ServerResolverBuilder) ServerForAddr(addr string) (*metadata.Server, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	for _, server := range s.servers {
+		if server.Addr.String() == addr {
+			return server, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to find Consul server for address %q", addr)
+}
+
+// Build returns a new serverResolver for the given ClientConn. The resolver
+// will keep the ClientConn's state updated based on updates from Serf.
+func (s *ServerResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, _ resolver.BuildOption) (resolver.Resolver, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// If there's already a resolver for this connection, return it.
+	// TODO(streaming): how would this happen since we already cache connections in ClientConnPool?
+	if resolver, ok := s.resolvers[cc]; ok {
+		return resolver, nil
+	}
```

Comment:
Another question, I'm not sure when `Build` would be called again for the same ClientConn. Looking at the source `Build` appears to only be called once when `DialContext` is called, so I suspect this does not do anything.

Reply:
I'm not super clear on when gRPC itself calls into these interface methods. I think the docs are limited and this was build based on example code so I'm not 100% sure if it's necessary or not but seems OK to leave in?

Reply:
Ya, seems fine to leave it with the comment calling out that it may not actually be used.